src/stdlib/strings.bpf.c

```diff
@@ -218,8 +75,31 @@ m_str* __strerror(int errno, m_arg *out) {
   if (errno >= 0 && errno <= EHWPOISON) {
     result = &errors[errno];
   } else {
-    result = &unknown;
+    result = &unknown_error;
   }
   __builtin_memcpy(&out->data, result, sizeof(*out));
   return &out->data;
 }
+
+m_str* __syscall_name(int n, m_arg *out) {
+  m_str *result;
+  char ch;
+
+  if (n >= 0 && n < syscall_names_size) {
```

Comment:
Per Viktor's comment, a common problem is the verifier can't always chase constraints through intermediate variables. After this if statement, it has a register that is effectively a phi, and it can't realistically track through the full graph of constraints.

In general, the more "verifier-friendly" way to write this code is to eliminate the temporary pointers, and just split the control flow rather than using intermediate state.

For example, perhaps some structure like this would work:

```
if (n >= 0 && n < nr_syscalls) {
  if (syscalls[n][0] == 0) {
    memcpy(out, &unknown_syscall);
  } else {
    memcpy(out, &syscalls[n]);
  }
} else {
  memcpy(out, &unknown_syscall);
}
```

Reply:
Thanks, i try  it now.

Reply:
And this method could drop the helper `bpf_probe_read_kernel`.